### PR TITLE
Allow for partial transcoding to a DirectPlayProfile

### LINKED ISSUE #6149: [bug] Unnecessarily transcoding vp9 in Firefox
**Describe the bug**
Jellyfin on web client is transcoding during playback of vp9 video when vp9 is supported by Firefox and my computer. 

**System (please complete the following information):**
 - OS: Ubuntu Server 20.04.2
 - Virtualization: Docker
 - Clients: Browser
 - Browser: Firefox 88.0.1 (on Ubuntu 21.04)
 - Jellyfin Version: 10.7.5
 - Playback: Transcode
 - Installed Plugins: none? manually installed none, default plugins (AudioDB, MusicBrainz, OMDb)
 - Reverse Proxy: caddy
 - Base URL: none
 - Networking: Host
 - Storage: local

**To Reproduce**

1. Play vp9 video on firefox on ubuntu. Open playback information.

**Expected behavior**
Transcoding only audio, or nothing (idk if aac is supported on firefox)

**Server Logs**

**FFmpeg Logs**

**Browser Console Logs**

**Screenshots**

Screenshot from 2021-06-04 03-52-16 censored

**Additional context**
Video was downloaded with youtube-dl, not sure how it handles metadata if that is relevant? Playback was also occasionally dropping frames, I assume transcoding may have something to do with it.

If you need me to add server logs, ffmpeg logs, or browser console logs, I am willing, but do not know how.

### LINKED ISSUE #6450: Avoid unnecessary transcoding by only transcoding incompatible streams
**Describe the bug**
This is a meta-bug that causes #6149, #6449 (TODO: look for other bugs relating to this).

If only some of the streams (or the container itself) is unsupported, only transcode the unsupported streams and just copy/mux the other streams.

**System (please complete the following information):**
 - OS: All
 - Virtualization: Any
 - Clients: All
 - Browser: All
 - Jellyfin Version: 10.7.6
 - Playback: Transcode
 - Hardware Acceleration: All
 - Installed Plugins: All
 - Reverse Proxy: All
 - Base URL: All
 - Networking: All
 - Storage: All

**To Reproduce**

1. Attempt to play a video to a client that supports the video codec but not the audio or container. For example, play a (MKV, VP9, AAC) video on Firefox.
2. Notice that transcoding occurs unnecessarily for the video stream as well as the audio stream.

**Expected behavior**

Only the unsupported the streams should be transcoded and supported streams should be passed through.

**Additional context**

This exposes a deeper issue with some of the transcoding logic that needs to be fixed.

Refer to `MediaBrowser.Model.Dlna.Streambuilder.BuildVideoItem()` for a starting reference point for this bug.

For example, something similar to this logic should be used to pick a target transcode output format:

```csharp
// Can't direct play, find the transcoding profile
TranscodingProfile transcodingProfile = null;
int bestMaxAudioChannels = 0;
foreach (var i in options.Profile.TranscodingProfiles)
{
    if (i.Type == playlistItem.MediaType && i.Context == options.Context)
    {
        int curMaxAudioChannels = 0;
        int.TryParse(i.MaxAudioChannels, out curMaxAudioChannels);

        if (transcodingProfile == null)
        {
            transcodingProfile = i;
        }
        else if (i.VideoCodec == NormalizeSourceCodecToTranscodeCodec(item.VideoStream.Codec) && transcodingProfile.VideoCodec != NormalizeSourceCodecToTranscodeCodec(item.VideoStream.Codec))
        {
            // Container supports original codec, so pick this profile instead.
            transcodingProfile = i;
        }
        else if (curMaxAudioChannels > bestMaxAudioChannels)
        {
            // Container supports more audio channels output.
            transcodingProfile = i;
        }
        // Probably add logic about containers that support embedded subtitles.
    }
}
```

Then, there should be some indication, perhaps in `MediaBrowser.Model.Dlna.StreamInfo`, to indicate whether to pass through the audio or video streams to the output container, and the transcoding logic should selectively transcode.

### LINKED ISSUE #7295: [Issue]: Audio-only transcoding with ffmpeg loses HDR information
### Please describe your bug

Given a media file with a HVEC video track in a mkv container and an AC3 default audio track, my device is capable of direct play and will switch into HDR mode.

However, if I switch the audio track to something that my device can't play, which in this case is a DTS audio track, transcoding of just the audio track needs to happen, which in this case it does with ffmpeg. Now my device will no longer switch into HDR mode, since the information in the mkv container about it being HDR seems to get lost in the process of transcoding with ffmpeg.

I've read a bit on the web and it seems that there's some combination of parameters that one can pass to ffmpeg to preserve the HDR info bit, but I'm no expert and not sure which one actually works.

If I can help debugging further, please let me know.

Originally reported in https://github.com/jellyfin/jellyfin-webos/issues/76

### Jellyfin Version

10.7.7

### if other:

_No response_

### Environment

```markdown
- OS: Ubuntu Server
- Virtualization: Docker
- Clients: LG 49UN71006LB
- Browser: Firefox
- FFmpeg Version: Docker
- Playback Method: Direct Play / Transcode
- Hardware Acceleration: None
- Plugins: None
- Reverse Proxy: None
- Base URL: None
- Networking: Host
- Storage: local
```

### Jellyfin logs

_No response_

### FFmpeg logs

_No response_

### Please attach any browser or client logs here

_No response_

### Please attach any screenshots here

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by GodTamIt**
**Changes**
This change allows for partial transcodes to be prioritized over full transcodes when a video stream can be decoded by the target client. In an instance where a video stream is fully compliant with the specifications of the client but transcoding is still required due to container being unsupported or other streams needing transcoding, this is where this change will prefer to target that direct profile instead of a full transcode to a `TranscodeProfile`.

**Note:** this change currently requires the resulting protocol to be HTTP progressive download. There is a `TODO` comment left there to detect whether the client may support something like HLS or DASH. I will say a codec-agnostic adaptive format like DASH seems kinda nice here.

Currently testing on Firefox, Chrome, Android TV, and Android.

**Issues**
Fixes #6149 
Fixes #6450
Fixes #7295

### CODE REVIEW

**Comment by cvium** [comment]
@GodTamIt Looks like we introduced some merge conflicts. Can you fix them?

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 1]
```diff
@@ -299,50 +299,10 @@ private StreamInfo BuildAudioItem(MediaSourceInfo item, AudioOptions options)
 
             if (directPlayMethods.Any())
             {
-                string audioCodec = audioStream?.Codec;
+                var conditions = GetConditionsForCodec(audioStream?.Codec, item.Container, options.Profile.CodecProfiles, CodecType.Audio, item, audioStream);
+                var failedConditions = FindFailedConditions(conditions, CodecType.Audio, item, audioStream);
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:303]: You don't seem to use the result of `FindFailedConditions`, so wouldn't it make more sense to change the return type to "bool" like "AllConditionsSatisfied"?

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 2]
```diff
@@ -665,251 +587,409 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
-            bool isEligibleForDirectStream = options.EnableDirectStream && (options.ForceDirectStream || directStreamEligibilityResult.Item1);
-
-            _logger.LogInformation(
-                "Profile: {0}, Path: {1}, isEligibleForDirectPlay: {2}, isEligibleForDirectStream: {3}",
-                options.Profile.Name ?? "Unknown Profile",
-                item.Path ?? "Unknown path",
-                isEligibleForDirectPlay,
-                isEligibleForDirectStream);
+            // First, try to DirectPlay.
+            var (canDirectPlayStream, transcodeReasons) = TryBuildVideoStreamForDirectPlay(item, options, videoStream, audioStream, subtitleStream, outputStream);
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:602]: I'm not a fan of having `transcodeReasons` be returned by one method and then altered in other methods. Either they should all return `transcodeReasons` or receive it as input (also not a big fan of side effects, but for private methods it's ok)

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 3]
```diff
@@ -665,251 +587,409 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
                     {
-                        var subtitleProfile = GetSubtitleProfile(item, subtitleStream, options.Profile.SubtitleProfiles, directPlay.Value, _transcoderSupport, item.Container, null);
-
-                        playlistItem.SubtitleDeliveryMethod = subtitleProfile.Method;
-                        playlistItem.SubtitleFormat = subtitleProfile.Format;
+                        return 128000;
                     }
 
-                    return playlistItem;
+                    return (audioChannels ?? 0) >= 6 ? 640000 : 384000;
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:640]: This should be the same
```suggestion
                    return audioChannels >= 6 ? 640000 : 384000;
```

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 4]
```diff
@@ -665,251 +587,409 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
+                {
+                    if ((audioChannels ?? 0) < 2)
+                    {
+                        return 768000;
+                    }
 
-            if (directPlayEligibilityResult.Item2.HasValue)
-            {
-                transcodeReasons.Add(directPlayEligibilityResult.Item2.Value);
+                    return (audioChannels ?? 0) >= 6 ? 3584000 : 1536000;
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:651]:
```suggestion
                    return audioChannels >= 6 ? 3584000 : 1536000;
```

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 5]
```diff
@@ -665,251 +587,409 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
+            }
 
-                playlistItem.PlayMethod = PlayMethod.Transcode;
+            yield break;
+        }
 
-                SetStreamInfoOptionsFromTranscodingProfile(playlistItem, transcodingProfile);
+        private IEnumerable<ProfileCondition> FindFailedConditions(IEnumerable<ProfileCondition> conditions, CodecType codecType, MediaSourceInfo mediaSource, MediaStream stream)
+        {
+            var failedConditions = new List<ProfileCondition>();
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:689]: Isn't this unused?

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 6]
```diff
@@ -1041,159 +1121,56 @@ private static int GetMaxAudioBitrateForTotalBitrate(long totalBitrate)
...
-                        foreach (ProfileCondition applyCondition in i.ApplyConditions)
-                        {
-                            if (!ConditionProcessor.IsVideoAudioConditionSatisfied(applyCondition, audioChannels, audioBitrate, audioSampleRate, audioBitDepth, audioProfile, isSecondaryAudio))
-                            {
-                                // LogConditionFailure(profile, "VideoAudioCodecProfile.ApplyConditions", applyCondition, mediaSource);
-                                applyConditions = false;
-                                break;
-                            }
-                        }
+                transcodeReasons.UnionWith(GetAllTranscodeReasonsForFailedConditions(failedContainerConditions, profile, "VideoContainerProfile", mediaSource));
```
Comment (cvium) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:1159]: I think the streambuilder is actually new'd everytime it's used, so `transcodeReasons` could just be a private field.

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 7]
```diff
@@ -655,7 +577,7 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
                 throw new ArgumentNullException(nameof(item));
             }
 
-            StreamInfo playlistItem = new StreamInfo
+            StreamInfo outputStream = new StreamInfo
```
Comment (crobibero) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:580]:
```suggestion
            StreamInfo outputStreamInfo = new StreamInfo
```

This object isn't a `Stream`

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 8]
```diff
@@ -1041,159 +1118,56 @@ private static int GetMaxAudioBitrateForTotalBitrate(long totalBitrate)
...
-
-            TransportStreamTimestamp? timestamp = videoStream == null ? TransportStreamTimestamp.None : mediaSource.Timestamp;
-            int? packetLength = videoStream?.PacketLength;
-            int? refFrames = videoStream?.RefFrames;
+            var failedContainerConditions = FindFailedConditions(containerConditions, CodecType.Video, mediaSource, videoStream);
 
-            int? numAudioStreams = mediaSource.GetStreamCount(MediaStreamType.Audio);
-            int? numVideoStreams = mediaSource.GetStreamCount(MediaStreamType.Video);
+            // Check video conditions
+            var videoCodecConditions = GetConditionsForCodec(videoStream!.Codec, mediaSource.Container, profile.CodecProfiles, CodecType.Video, mediaSource, videoStream);
```
Comment (crobibero) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:1141]:
```suggestion
            var videoCodecConditions = GetConditionsForCodec(videoStream.Codec, mediaSource.Container, profile.CodecProfiles, CodecType.Video, mediaSource, videoStream);
```

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 9]
```diff
@@ -1041,159 +1118,56 @@ private static int GetMaxAudioBitrateForTotalBitrate(long totalBitrate)
...
-                    LogConditionFailure(profile, "VideoCodecProfile", i, mediaSource);
-
-                    var transcodeReason = GetTranscodeReasonForFailedCondition(i);
-                    var transcodeReasons = transcodeReason.HasValue
-                        ? new List<TranscodeReason> { transcodeReason.Value }
-                        : new List<TranscodeReason>();
-
-                    return (null, transcodeReasons);
-                }
+                var audioCodecConditions = GetConditionsForCodec(videoStream!.Codec, mediaSource.Container, profile.CodecProfiles, CodecType.VideoAudio, mediaSource, audioStream);
```
Comment (crobibero) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:1148]:
```suggestion
                var audioCodecConditions = GetConditionsForCodec(videoStream.Codec, mediaSource.Container, profile.CodecProfiles, CodecType.VideoAudio, mediaSource, audioStream);
```

File: MediaBrowser.Model/Dlna/StreamBuilder.cs (hunk 7 above)
Comment (GodTamIt) [reply MediaBrowser.Model/Dlna/StreamBuilder.cs]: Addressed same naming convention in other functions as well :+1:

File: MediaBrowser.Model/Dlna/StreamBuilder.cs [hunk 10]
```diff
@@ -665,251 +582,405 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
...
                             {
-                                // LogConditionFailure(options.Profile, "VideoCodecProfile.ApplyConditions", applyCondition, item);
-                                applyConditions = false;
-                                break;
+                                yield return condition;
                             }
+                            break;
                         }
+                    case CodecType.VideoAudio:
+                        {
```
Comment (crobibero) [inline MediaBrowser.Model/Dlna/StreamBuilder.cs:706]:
```suggestion
                    case CodecType.VideoAudio:
                    {
```

the indentation for these `case` statements is off,
and "Closing brace should be followed by blank line" as the GH analyzer is reporting

Comment (GodTamIt) [reply MediaBrowser.Model/Dlna/StreamBuilder.cs]: If you look at other parts of the file, the `case` braces look exactly like this. This is also how the formatter formats the braces.

I did fix the closing brace blank line warning though.

**Comment by crobibero** [comment]
@GodTamIt We would like to get this into 10.8; if you don't have availability to complete this PR can you please allow edits from maintainers?

**Comment by nyanmisaka** [comment]
Does fast forward work fine with Progressive playback?

And how about the latency and network traffic comparing with HLS?

**Comment by GodTamIt** [comment]
> Does fast forward work fine with Progressive playback?

Yes, it seems to work just fine for me.

**Review by crobibero — APPROVED** [review]

**Comment by nyanmisaka** [comment]
I just tried this with Chrome. It throw me an error on webm file which contains one VP9 video stream and one OPUS audio stream.

Something went wrong in ConditionProcessor.
```
[ERR] [23] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL POST /Items/a5b48fc4b9657c120305a66e6fdfd4dc/PlaybackInfo.
System.ArgumentException: Unexpected condition on audio file: IsSecondaryAudio
   at MediaBrowser.Model.Dlna.ConditionProcessor.IsAudioConditionSatisfied(ProfileCondition condition, Nullable`1 audioChannels, Nullable`1 audioBitrate, Nullable`1 audioSampleRate, Nullable`1 audioBitDepth)
   at MediaBrowser.Model.Dlna.StreamBuilder.FindFailedConditions(IEnumerable`1 conditions, CodecType codecType, MediaSourceInfo mediaSource, MediaStream stream)+MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)
   at MediaBrowser.Model.Dlna.StreamBuilder.GetVideoDirectPlayProfile(VideoOptions options, MediaSourceInfo mediaSource, MediaStream videoStream, MediaStream audioStream, Boolean isEligibleForDirectStream)
   at MediaBrowser.Model.Dlna.StreamBuilder.TryBuildVideoStreamForDirectPlay(MediaSourceInfo mediaSource, VideoOptions options, MediaStream videoStream, MediaStream audioStream, MediaStream subtitleStream, StreamInfo outputStreamInfo, HashSet`1 transcodeReasons)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(MediaSourceInfo item, VideoOptions options)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(VideoOptions options)
   at Jellyfin.Api.Helpers.MediaInfoHelper.SetDeviceSpecificData(BaseItem item, MediaSourceInfo mediaSource, DeviceProfile profile, AuthorizationInfo auth, Nullable`1 maxBitrate, Int64 startTimeTicks, String mediaSourceId, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String playSessionId, Guid userId, Boolean enableDirectPlay, Boolean enableDirectStream, Boolean enableTranscoding, Boolean allowVideoStreamCopy, Boolean allowAudioStreamCopy, IPAddress ipAddress)
   at Jellyfin.Api.Controllers.MediaInfoController.GetPostedPlaybackInfo(Guid itemId, Nullable`1 userId, Nullable`1 maxStreamingBitrate, Nullable`1 startTimeTicks, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String mediaSourceId, String liveStreamId, Nullable`1 autoOpenLiveStream, Nullable`1 enableDirectPlay, Nullable`1 enableDirectStream, Nullable`1 enableTranscoding, Nullable`1 allowVideoStreamCopy, Nullable`1 allowAudioStreamCopy, PlaybackInfoDto playbackInfoDto)
   at lambda_method847(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
```

**Comment by GodTamIt** [comment]
> I just tried this with Chrome. It throw me an error on webm file which contains one VP9 video stream and one OPUS audio stream.
> 
> Something went wrong in ConditionProcessor.
> 
> ```
> [ERR] [23] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL POST /Items/a5b48fc4b9657c120305a66e6fdfd4dc/PlaybackInfo.
> System.ArgumentException: Unexpected condition on audio file: IsSecondaryAudio
>    at MediaBrowser.Model.Dlna.ConditionProcessor.IsAudioConditionSatisfied(ProfileCondition condition, Nullable`1 audioChannels, Nullable`1 audioBitrate, Nullable`1 audioSampleRate, Nullable`1 audioBitDepth)
>    at MediaBrowser.Model.Dlna.StreamBuilder.FindFailedConditions(IEnumerable`1 conditions, CodecType codecType, MediaSourceInfo mediaSource, MediaStream stream)+MoveNext()
>    at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)
>    at MediaBrowser.Model.Dlna.StreamBuilder.GetVideoDirectPlayProfile(VideoOptions options, MediaSourceInfo mediaSource, MediaStream videoStream, MediaStream audioStream, Boolean isEligibleForDirectStream)
>    at MediaBrowser.Model.Dlna.StreamBuilder.TryBuildVideoStreamForDirectPlay(MediaSourceInfo mediaSource, VideoOptions options, MediaStream videoStream, MediaStream audioStream, MediaStream subtitleStream, StreamInfo outputStreamInfo, HashSet`1 transcodeReasons)
>    at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(MediaSourceInfo item, VideoOptions options)
>    at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(VideoOptions options)
>    at Jellyfin.Api.Helpers.MediaInfoHelper.SetDeviceSpecificData(BaseItem item, MediaSourceInfo mediaSource, DeviceProfile profile, AuthorizationInfo auth, Nullable`1 maxBitrate, Int64 startTimeTicks, String mediaSourceId, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String playSessionId, Guid userId, Boolean enableDirectPlay, Boolean enableDirectStream, Boolean enableTranscoding, Boolean allowVideoStreamCopy, Boolean allowAudioStreamCopy, IPAddress ipAddress)
>    at Jellyfin.Api.Controllers.MediaInfoController.GetPostedPlaybackInfo(Guid itemId, Nullable`1 userId, Nullable`1 maxStreamingBitrate, Nullable`1 startTimeTicks, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String mediaSourceId, String liveStreamId, Nullable`1 autoOpenLiveStream, Nullable`1 enableDirectPlay, Nullable`1 enableDirectStream, Nullable`1 enableTranscoding, Nullable`1 allowVideoStreamCopy, Nullable`1 allowAudioStreamCopy, PlaybackInfoDto playbackInfoDto)
>    at lambda_method847(Closure , Object )
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
>    at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
> ```

Try again with the latest of the branch? I accidentally flipped two conditions on a refactor.

**Comment by nyanmisaka** [comment]
> Try again with the latest of the branch? I accidentally flipped two conditions on a refactor.

It works this time.

Do you happened to have a video sample that can trigger Progressive playback in Chrome or Firefox?

**Comment by justinas** [comment]
Tried this branch with the Android TV client (0.12.3), seems like in some cases the output file name ends up being invalid:


```
[2022-01-09 19:27:55.920 +00:00] [ERR] [57] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL "GET" "/videos/2b367999-ace0-2881-a456-1f3bc8d06fe9/stream.m4v,mov,xvid,vob,mkv,wmv,asf,ogm,ogv,mp4,webm".
MediaBrowser.Common.FfmpegException: FFmpeg exited with code 1
```

ffmpeg:
```
<...>
[NULL @ 0x22bb500] Unable to find a suitable output format for '/var/lib/jellyfin/transcodes/9d786a94a315069a90fbb0fe68bd0ea4.m4v,mov,xvid,vob,mkv,wmv,asf,ogm,ogv,mp4,webm'
/var/lib/jellyfin/transcodes/9d786a94a315069a90fbb0fe68bd0ea4.m4v,mov,xvid,vob,mkv,wmv,asf,ogm,ogv,mp4,webm: Invalid argument
```

**Comment by GodTamIt** [comment]
> Do you happened to have a video sample that can trigger partial transcoding (Progressive) in Chrome or Firefox?

@nyanmisaka I believe the attached should do the trick. It's VP9 + AAC, which the browser won't like.

It's super crummy quality but should do the trick for testing. Side note: it's in an mp4 extension but it's actually an mkv container.


https://user-images.githubusercontent.com/3161130/148802022-0ee0cb79-f944-4f10-b2f9-c81255819e62.mp4

**Comment by GodTamIt** [comment]
> Tried this branch with the Android TV client (0.12.3), seems like in some cases the output file name ends up being invalid:
> 
> ```
> [2022-01-09 19:27:55.920 +00:00] [ERR] [57] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL "GET" "/videos/2b367999-ace0-2881-a456-1f3bc8d06fe9/stream.m4v,mov,xvid,vob,mkv,wmv,asf,ogm,ogv,mp4,webm".
> MediaBrowser.Common.FfmpegException: FFmpeg exited with code 1
> ```
> 
> ffmpeg:
> 
> ```
> <...>
> [NULL @ 0x22bb500] Unable to find a suitable output format for '/var/lib/jellyfin/transcodes/9d786a94a315069a90fbb0fe68bd0ea4.m4v,mov,xvid,vob,mkv,wmv,asf,ogm,ogv,mp4,webm'
> /var/lib/jellyfin/transcodes/9d786a94a315069a90fbb0fe68bd0ea4.m4v,mov,xvid,vob,mkv,wmv,asf,ogm,ogv,mp4,webm: Invalid argument
> ```

I'm assuming this works in the mainline? Do you have any files that could help me reproduce this?

**Comment by justinas** [comment]
@GodTamIt 

[This demo](https://www.demolandia.net/downloads.html?id=25195020) seems to reproduce it for me.

```
Input #0, matroska,webm, from 'dolby-city-redux-lossless-(www.demolandia.net).mkv':
  Metadata:
    encoder         : libebml v1.3.0 + libmatroska v1.4.1
    creation_time   : 2015-04-16T18:58:38.000000Z
  Duration: 00:00:24.02, start: 0.000000, bitrate: 18675 kb/s
  Stream #0:0(eng): Video: h264 (High), yuv420p(progressive), 1920x1080, SAR 1:1 DAR 16:9, 29.97 fps, 29.97 tbr, 1k tbn, 59.94 tbc (default)
  Stream #0:1(eng): Audio: truehd, 48000 Hz, 7.1, s32 (24 bit) (default)
```

I would expect it to transcode TRUEHD as ExoPlayer as it exists in jellyfin-androidtv does not support it.

**Comment by Shadowghost** [comment]
What's the state on this? Would be great if we can get that an the other PRs depending on it in before 10.8 feature freeze hits.

**Comment by justinas** [comment]
> @justinas I just tried both the `copyright-free-music.mkv` and the `dolby-city-redux-lossless` on my local build. Couldn't reproduce your issues that you mentioned

@GodTamIt interesting. I will try this again at some point and create a new issue if need be. As it is not reproducible by others, please do not consider it a blocker.

**Review by crobibero — APPROVED** [review]

**Comment by eyezak** [comment]
I took a bit of a different approach with #7325 which I think will support these use cases with transcode profiles and HLS streaming

**Comment by crobibero** [comment]
Thank you for your contribution. We went with [#7325](https://github.com/jellyfin/jellyfin/pull/7325) as it included a bunch of tests.

Closing in favor of https://github.com/jellyfin/jellyfin/pull/7325

**Pull request closed without merging**